replica/database.cc

```diff
@@ -494,7 +509,7 @@ backlog_controller::control_point compaction_controller::last_control_point(floa
     //      x2 = (y2 – y1) / m + x1
     // For finding x2 (or normalization factor), we replace m by
     // the constant slope with default max shares (1000).
-    float nf = (max_shares - middle_point().output) / base_slope + middle_point().input;
+    float nf = (max_shares - middle_point().output) / (af * base_slope) + middle_point().input;
 
```

Comment:
Again I'm confused by the calculation. Shouldn't we have shares *= af or something?

btw, we should find a more descriptive and less violent name for it.

Reply:
This is calculating normalization factor (NF), not shares

The idea was to abstract NF as an "aggressive factor" (AF), making it easier to reason about and not expose an implementation detail.

Slope formula:
```
    m = (y2 – y1) / (x2 – x1)
    
    where x2 = NF and y2 = max_shares.
```

To find NF given max shares (while preserving slope):
```
    x2 = (y2 – y1) / m + x1

    where m is the slope we're targetting (referred internally as base slope).
```

That's why you see:
```
    float nf = (max_shares - middle_point().output) / base_slope + middle_point().input;
```

The higher the NF the lower the aggressiveness.

Therefore it's natural that we apply AF to the "base_slope", resulting in
```
    float nf = (max_shares - middle_point().output) / (af * base_slope) + middle_point().input;
```

That controls the steepness of the slope. The higher the AF, the lower the NF, therefore the higher the aggressiveness.

This is what it looks like in practice:

```
            Aggressive factor
Backlog     0.5         1.0         2.0
         |=================================
1        |  83.3333     83.3333     83.3333
2        |  107.895     115.789     131.579
4        |  139.474     178.947     257.895
8        |  202.632     305.263     510.526
16       |  328.947     557.895     1000
32       |  581.579     1000        1000
64       |  1000        1000        1000
```

![backlog_out](https://github.com/scylladb/scylladb/assets/1409139/129aed91-0c6c-4041-9126-11adba6a7613)

So the aggressive name makes sense, but I agree it sounds scary. The opposite would be "sweetness" factor, but the name might be so welcoming that one might bump it while not being aware of the consequences :-)

Reply:
Backlog-to-shares formula is:

`    y = y1 + (x - x1) * ((y2 - y1) / (x2 - x1))`
    
1) y2 is NF

2) x is the normalized backlog
    
3) we can think of slope (((y2 - y1) / (x2 - x1))) as backlog weight 

so perhaps aggressive factor can be renamed to compaction_backlog_weight_factor, or simply compaction_backlog_weight, default'ing to 1.0f.

FYI @bhalevy

Reply:
In audio engineering this is called called "gain".

https://www.antarestech.com/community/whats-the-difference-between-gain-and-volume
> It is the ratio between the input volume and the output volume

Reply:
> I think it should be decoupled from the linear interpolation calculation, it's mixing two things there.
> 
> We can just make normlization_factor a tunable, no?

The idea of abstractin normalization factor came after a few attempts to explain it to others. Benny and I were concerned nobody besides a few would understand it. 

If we come to think of it, the idea of "backlog weight" ((y2 - y1) / (x2 - x1)) between two points is mathematically implicit in the interface. compaction_controller defines all of its control points. That's the only part of formula that we play with. If you have two control points, (x1, y1) and (x2, y2), where x2 is NF, then what we do is to pick a NF, given the supplied params (max shares and sensitivity), that will yield a target backlog weight.

Reply:
@bhalevy, @avikivity: I see value in exposing normalization_factor as is (predictability is one), but then its usage would be restricted to a few experts. If I were to implement two tunables, without any complexity derived from playing with controls points, then we'd have something like: compaction-max-shares (but then I'd patch controller to remove 1k limit, and do std::min(compaction-max-shares, controller_output), so we don't have to play with normalization factor to preserve slope), and compaction-normalization-factor (good doc would need to come along).

Reply:
I agree with Raphael.  Exposing `normalization_factor` as is might be simpler at the coding level, but it's much harder to work with from the user's perspective. On the other hand, an abstract "backlog sensitivity" tunable is much simpler to comprehand and the model provides more sensible semantics towards the user, with the low penalty of a bit of algebra under the hood.

Reply:
If the weight was in the range 0-1, it would make sense, but it's really 0-inf with 1 being the current default.

So why not keep it 0-inf with 30 being the default? It means one doesn't need to think about the hidden 30. We can later replace 30 by the disk-to-ram ratio (if we think it's a good idea, possibly it isn't).